### PR TITLE
Filter development listings

### PR DESCRIPTION
**Pull request by pmargreff**
Allow the client to filter listings while querying from developments as root. It doesn't accept pagination. 

Also made some minor refactor extracting `has_admin_rights` to Listings context once it is used in different places and rewrite developments `has_admin_rights` to look only to user role (once only admin has admin rights on this contexts).

### CODE REVIEW

**Comment by rhnonose** [comment]
LGTM 👍 
But I think if we're refactoring `has_admin_rights?`, we should put in context's `Policy` module.
This was a lazy workaround because our `Policy` logic didn't return a boolean, but permissions should be there.

File: apps/re/test/listings/listings_test.exs [hunk 1]
```diff
@@ -695,6 +695,27 @@ defmodule Re.ListingsTest do
...
+  describe "has_admin_rights?/2" do
+    test "is true when user is admin" do
+      listing = build(:listing)
+      user = %{role: "admin"}
+      assert Listings.has_admin_rights?(listing, user) == true
+    end
+
+    test "is true when user is owner" do
+      user = insert(:user)
+      listing = insert(:listing, user: user)
```
Comment (joaodubas) [inline apps/re/test/listings/listings_test.exs]: Maybe you can just populate the Listing and the User:

```elixir
user = Re.User{id: 1}
listing = Re.Listing{user_id: 1}
```

Comment (joaodubas) [inline apps/re/test/listings/listings_test.exs]: Maybe you can just populate the Listing and the User:

```elixir
user = Re.User{id: 1}
listing = Re.Listing{user_id: 1}
```

**Comment by pmargreff** [comment]
> But I think if we're refactoring has_admin_rights?, we should put in context's Policy module. 

Makes sense, I'll change it.

File: apps/re_web/lib/graphql/resolvers/addresses.ex [hunk 2]
```diff
@@ -6,19 +6,26 @@ defmodule ReWeb.Resolvers.Addresses do
...
-    admin? = admin?(current_user)
+  def per_development(development, params, %{context: %{current_user: current_user}}) do
+    admin = access_complete_address?(current_user, development)
 
-    {:address, Map.put(params, :has_admin_rights, admin?)}
+    {:address, Map.put(params, :has_admin_rights, admin)}
+  end
+
+  defp access_complete_address?(user, parent) do
+    case Policy.authorize(:show_complete_address, user, parent) do
```
Comment (rhnonose) [inline apps/re_web/lib/graphql/resolvers/addresses.ex]: You shouldn't call `Policy` directly, should should go through the context.

File: apps/re_web/lib/graphql/resolvers/images.ex [hunk 3]
```diff
@@ -121,8 +123,12 @@ defmodule ReWeb.Resolvers.Images do
   def insert_image_trigger(%{parent: %Re.Development{uuid: uuid}}),
     do: "development_updated:#{uuid}"
 
-  defp admin?(%{role: "admin"}), do: true
-  defp admin?(_), do: false
+  defp access_inactive_images?(user, parent) do
+    case Policy.authorize(:show_inactive_images, user, parent) do
```
Comment (rhnonose) [inline apps/re_web/lib/graphql/resolvers/images.ex]: You shouldn't call Policy directly, should should go through the context.

File: apps/re_web/lib/graphql/resolvers/addresses.ex (hunk 2 above)
Comment (pmargreff) [reply apps/re_web/lib/graphql/resolvers/addresses.ex]: yes, for some reason the version you reviewed is outdated, can you go to the last one?

File: apps/re_web/lib/graphql/resolvers/images.ex [hunk 4]
```diff
@@ -121,16 +123,26 @@ defmodule ReWeb.Resolvers.Images do
   def insert_image_trigger(%{parent: %Re.Development{uuid: uuid}}),
     do: "development_updated:#{uuid}"
 
+  defp has_admin_rights?(user, %Listing{} = listing) do
+    case Listings.authorize(:has_admin_rights, user, listing) do
```
Comment (rhnonose) [inline apps/re_web/lib/graphql/resolvers/images.ex]: My apologies, I wasn't clear with this one.
It should be used like described in `bodyguard`.
Example: `Bodyguard.permit(Listings, :edit_listing, user, listing)`.
It uses the `Bodyguard.Policy` behaviour.

Comment (pmargreff) [reply apps/re_web/lib/graphql/resolvers/images.ex]: I got you now, I didn't know it is the right way to do it. 
Thank you for the explanation.

Comment (pmargreff) [reply apps/re_web/lib/graphql/resolvers/images.ex]: Also [saw in the documentation ](https://hexdocs.pm/bodyguard/Bodyguard.html#permit?/4)that `permit?` does convert the response to bool (:ok -> true, _ -> false) for free.

Comment (rhnonose) [reply apps/re_web/lib/graphql/resolvers/images.ex]: That's a great finding, didn't know that 🤔

**Pull request merged**